### DISCUSSION TITLE
Improve developer experience for VSCode

### DISCUSSION THREAD

**Discussion by foxtrotcz** [craftcms/cms#13600] [Ideas]
Hello,
VSCode is the most popular IDE.
74% usage overall, [StackOverflow Survey](https://survey.stackoverflow.co/2023/#most-popular-technologies-new-collab-tools)
57% in Craft ecosystem, [Honcho Survey](https://honcho.agency/craft-cms/craft-cms-developer-benchmark-results-2022#your-craft-cms-setup)

But sadly developer experience for Twig templates and therefore Craft is not very good.

Some extensions exist but they are often not updated or not reliable. [Link](https://marketplace.visualstudio.com/search?term=twig&target=VSCode&category=All%20categories&sortBy=Relevance)
Twig plugin for Prettier recently stopped working because of new Prettier version and is also not updated. [Link](https://github.com/trivago/prettier-plugin-twig-melody)
And autocomplete for Craft exists only for PHPStorm. [Link](https://github.com/nystudio107/craft-autocomplete)

I recently listened to interview with Brandon Kelly where he says that if they were to do some additional project/solution it will be only something where they can bring some significant additional value by doing it first-party. [Link](https://craftentries.io/podcast/craft-cms-ten-years-and-beyond?t=1%3A01%3A24)
Nitro is over and now we use DDEV so I guess that was example what is better not doing in-house but leave it to someone else.

So I wonder if improving developer experience by creating VSCode extension is something that would work better as in-house project.
[Antlers Toolbox](https://antlers.dev/) for Statamic can serve as an example.

Either way, I think any solution that would improve developer experience for Craft/Twig in VSCode would be very welcome by a lot of people.

Thanks.

**Comment by khalwat** [comment]
It probably would be best as a project funded by P&T (or a combination of P&T and Symfony or other interested parties) but written by a third party contractor with expertise in this area.

VS Code Language Server: https://code.visualstudio.com/api/language-extensions/language-server-extension-guide

**Comment by mandrasch** [comment]
👍 👏  A Twig plugin for Prettier would be very helpful to enable the same consistent auto formatting in both VSCode and PhpStorm (or other IDEs supporting prettier, git hooks / github actions with prettier). 

> my main painpoint is currently that we often copy & paste / move some code parts between twig templates and can't auto-indent them properly (in VS Code). this would already improve the developer experience a lot 🤓

The inactive [twig melody plugin](https://github.com/trivago/prettier-plugin-twig-melody) would need a migration to prettier v3 if it would be used for further development:
- https://github.com/prettier/prettier/wiki/How-to-migrate-my-plugin-to-support-Prettier-v3%3F

There are some newer community prettier plugins which could also be a reference:
- e.g. https://github.com/withastro/prettier-plugin-astro
- full list https://prettier.io/docs/en/plugins#community-plugins

btw: TYPO3 has the same challenge since they have no prettier plugin for Fluid since years, seems to be a challenge everywhere.

Unfortunately I have no experience with prettier plugins - an estimate of how much work would be necessary would helpful I guess? Maybe anyone in the community has experience with this?

**Comment by BigglesZX** [comment]
Just to add a +1 to this – even just a reliable Twig formatter (ideally as a plugin for Prettier) would be a huge step forward in DX in my opinion. Formatting JS and CSS/SCSS is trivial nowadays but the template side of things is being left behind :(

I'm currently using VSCode with [this Prettier plugin](https://github.com/wearedestination/prettier-plugin-twig), which is a fork of Shopify's Liquid formatter, and while it's very usable and some great work has gone into it, it has some rough edges, particularly around `if..elseif` and nested object syntax.

A canonical method of formatting Twig + HTML supported by the Craft developers would be amazing. If this were based on Prettier it would add value to a much wider audience, not just VSCode users.

**Reply by mandrasch** [reply]
👍 

jfyi - here is another fork, but the forking dev added warning "Not ready. DON'T USE!" recently
https://github.com/zackad/prettier-plugin-twig

Twiggy looked also promising, but no formatting yet / plans for prettier: https://github.com/moetelo/twiggy/discussions/24

**Comment by foxtrotcz** [comment]
Some new development, Symfony language tools for VSCode were introduced.
I wonder if they will someday provide good Twig tooling.
https://github.com/symfony/language-tools
